src/waypoint.jsx

```diff
@@ -333,13 +352,30 @@ export default class Waypoint extends React.Component {
    * @return {Object}
    */
   render() {
+    if (this.props.noWrapper) {
+      const child = React.Children.only(this.props.children);
+      const ref = (node) => {
+        this.refElement(node);
+        if (this.props.children.ref) {
+          this.props.children.ref(node);
+        }
+      };
+      return React.cloneElement(child, { ref });
+    }
+
+    if (this.props.children) {
+      return <div ref={this.refElement}>{this.props.children}</div>;
+    }
+
     // We need an element that we can locate in the DOM to determine where it is
     // rendered relative to the top of its context.
     return <span ref={this.refElement} style={{ fontSize: 0 }} />;
   }
 }
 
 Waypoint.propTypes = {
+  children: PropTypes.oneOfType([PropTypes.node, PropTypes.array]),
+  noWrapper: PropTypes.bool,
```

Comment:
The proposal from @trotzig sounds reasonable to me. My biggest concerns are: 

1) It'll be confusing that it behaves differently in different situations (but this should be solvable by looking at the docs). 
2) There may be times when you actually *want* the div inserted (though this is silly; the user could simply wrap a div themselves). 

> I think it would be much better to provide a useful error message and always require an only child that is not a functional component. That is clear, consistent, and flexible.

"Flexible" isn't quite the word I'd use for that 😉. Personally I'd be very surprised if an issue were not filed using a word more like "frustrating". 

My assumption here is that this library is intended for broad use, including by devs who don't understand React very well, and that convenience is an important consideration.  

In any case, @jmeas curious for your thoughts